src/dosext/mfs/mfs.c

```diff
@@ -1672,15 +1672,14 @@ struct mfs_dirent *dos_readdir(struct mfs_dir *dir)
       dir->de.d_name = dir->de.d_long_name = de->d_name;
     } else {
 #ifdef __linux__
-      struct kernel_dirent *de;
+      struct kernel_dirent de[2];
       int ret;
 
-      ret = (int)RPT_SYSCALL(ioctl(dir->fd, vfat_ioctl, (long)de));
+      ret = (int)RPT_SYSCALL(ioctl(dir->fd, vfat_ioctl, de));
       if (ret == -1 || de[0].d_reclen == 0)
-	return NULL;
-      de[0].d_name[min((size_t)de[0].d_reclen, sizeof(de[0].d_name)-1)] = '\0';
+        return NULL;
+
       dir->de.d_name = de[0].d_name;
-      de[1].d_name[min((size_t)de[1].d_reclen, sizeof(de[1].d_name)-1)] = '\0';
       dir->de.d_long_name = de[1].d_name;
```

Comment:
should we be copying the filenames somewhere, not setting pointers in mfs_dirent to local strings?

Reply:
or perhaps making it static would be enough?

Reply:
Yeah, pure crap!
I'd rather change `struct mfs_dirent`
to have arrays, similar to `struct dir_ent`
defined in the same header.
Note that `dos_opendir()` does
`ioctl(fd, vfat_ioctl, (long)&de)`
(extra ampersand)

Reply:
Yes I saw that. Casting always seems to cover up things which is why I try to avoid it. I think I probably need to write a test, but I don't really want to require root to be able to mount the vfat.

Reply:
I wonder how may people actually use vfat mounted linux filesystems?

Reply:
Is this what you'd be using if you mounted dos or windows fat/fat32 filesystem from DOSEMU ?
If so, I've definitely done this in the past.

Reply:
Yes, in the past everyone did that!
Like 20 years ago.
I don't think we should break it
intentionally, but I'd proceed with
any changes and wait for someone
to complain, if things do not work.
Obviously they didn't work anyway,
I would be very surprised if that
piece could ever work correctly.

Reply:
It seems that before 3a5150d3402d816939ae328bf94a83f22dd449e5 was added the structure was static. I can see how that might work, but it would seem this workaround cribbed from Wine was a tried and tested solution. I'll see if I can write a test.

Reply:
OK, whatever you think is best.

Reply:
It did work at the time I implemented it but that's quite a few years ago (it even exposed a kernel bug at the time)! I haven't tested dosemu on mounted vfat for years I must admit, but indeed this used to be standard practice.

Reply:
yes certainly 3a5150d is no longer relevant as almost nobody is running 2.6 kernels any more -- I still have access to some CentOS 6 servers with 2.6.32 but even that will be EOL very soon.

Reply:
I actually have removed it already in 316f667.
In any case, Andrew opened #1296 claiming
that vfat no longer works with LFNs.

Reply:
yes the issue with 316f667 is that a random garbage pointer in a register or on the stack is now passed to the ioctl. The same technique as before 3a5150d would fix #1296 I think.

Reply:
Now that I've used VFAT mounted under Linux with Dosemu I'm beginning to like the idea. I'm going to write some tests for it so at least once the LFNs get fixed it'll stay working.

Reply:
> The same technique as before 3a5150d would fix #1296 I think.

From what I saw in my somewhat hacked up local version, was that the code in lfn.c wasn't doing the vfat lookups itself, but it was asking the redirector instead and that does 8.3 lookups. If if were possible I'd prefer for the redirector to do the lookup (as now) but for LFN, and somehow augment the redirector interface to pass the LFN results back to lfn.c. I wonder if that's possible to do without changing the redirector interface, like @bartoldeman did with `extended open()`, but for sure there's some thinking to do around it.